### PR TITLE
deprecate resource names not supported by microgenerators

### LINKED ISSUE #3012: Python: Deprecate resource name helper functions
These functions will be going away in the microgenerator.

### PR DESCRIPTION
**Pull request by hzyi-google**
Deprecate resource names that will not exist in microgenerator/proto annotation world.
- add a `deprecated_collections` field in gapic config V2. Resource names specified here will continue to be generated the same way gapic generator treats `collections` in gapic config v1, but the helper functions will be marked deprecated.

- Add two deprecated resource names in baseline tests,  `project_book` to test the case where a deprecated resource name has the same name as a non-deprecated one, and `book_from_archive` to test the case where a deprecated resource has a unique name.

- Node.js change: add a deprecation warning to all resource name helpers related to `project_book` and `book_from_archive` ===> PTAL @alexander-fenster 

- Ruby: add a deprecation warning to all resource name helpers related to `book_from_archive` and `project_book` ===> PTAL @dazuma 

- Python: add a deprecation warning to **all** resource names. ===> PTAL @software-dov @busunkim96  (Fix #3012)

### CODE REVIEW

**Comment by software-dov** [comment]
Python generated warnings look good.

**Comment by hzyi-google** [comment]
@vam-google for implementation review.

File: src/main/resources/com/google/api/codegen/nodejs/main.snip [hunk 1]
```diff
@@ -369,6 +369,9 @@
 
 @private createResourceFunction(function, apiName)
   /**
+  @if function.isResourceNameDeprecated
+    {@""} * @@depreated Multi-pattern resource names will have unified formatting and parsing helper functions. This helper function will be deleted in the next major version.
```
Comment (alexander-fenster) [inline src/main/resources/com/google/api/codegen/nodejs/main.snip]: `depreated` should be `deprecated` :)

Comment (hzyi-google) [reply src/main/resources/com/google/api/codegen/nodejs/main.snip]: Whoops!

Comment (alexander-fenster) [reply src/main/resources/com/google/api/codegen/nodejs/main.snip]: LGTM otherwise.

**Comment by hzyi-google** [comment]
Hi @vam-google @dazuma 
I need to get this merged soon to catch an artman release (this is unrelated to the unsettled discussions for Java). Feel free to take a look and drop comments if you want to and I am happy to address any concerns in coming PRs.

File: src/main/resources/com/google/api/codegen/ruby/main.snip [hunk 2]
```diff
@@ -280,11 +280,18 @@
...
+    @# @@deprecated Multi-pattern resource names will have unified creation and parsing helper functions.
+    @# This helper function will be deleted in the next major version.
+  @end
   @join param : function.resourceIdParams
     @# @@param {@param.docName} [String]
   @end
   @# @@return [String]
   def self.{@function.name} {@createResourceFunctionParams(function.resourceIdParams)}
+    @if function.isResourceNameDeprecated
+      warn "This helper function will be deleted in the next major version."
```
Comment (dazuma) [inline src/main/resources/com/google/api/codegen/ruby/main.snip]: Please omit this. Ruby devs do not like warnings printed to stderr with no way of disabling it.

Comment (hzyi-google) [reply src/main/resources/com/google/api/codegen/ruby/main.snip]: Done.

**Comment by hzyi-google** [comment]
PTAL

**Pull request merged**